### PR TITLE
XSLT: x-sap-odm-entity-name and x-sap-odm-oid

### PR DESCRIPTION
**Pull request by ralfhandl**

### CODE REVIEW

File: tools/tests/annotations.openapi3.json [hunk 1]
```diff
@@ -2896,6 +2898,8 @@
             "Supported.Annotations.SinglePartKey-create": {
                 "type": "object",
                 "x-sap-object-node-type": "SontOne",
+                "x-sap-odm-entity-name": "odmOne",
+                "x-sap-odm-oid": "Computed",
```
Comment (pavelkornev) [inline tools/tests/annotations.openapi3.json:2902]: Should not `x-sap-odm-oid` reference to some existent property in the `properties` object? I don't see property `Computed`

Comment (ralfhandl) [reply tools/tests/annotations.openapi3.json:2902]: Good question: the OID is typically system-created, just as the `Computed` property in this example, so it is only present in the READ structure and not in the CREATE and UPDATE structures.

Should we restrict `x-sap-odm-oid` to the READ structure?

The `x-sap-odm-entity-name` makes sense in all three structures, so I'd keep it duplicated.

Comment (ralfhandl) [reply tools/tests/annotations.openapi3.json:2902]: @NilsObermiller-SAP, @swennemers

Comment (juergenheymann) [reply tools/tests/annotations.openapi3.json:2902]: I agree with Sebastian. I think the key points are:
* For most APIs, OID is an alternate key in addition to an existing local key. This means that it can happen that for all read/update/patch... operations either the local or the OID key can be given to identify the object.
  * For entities, the OID must be available as field in the DB (is also an S/4 rule) and must be exposed extraction interfaces etc. So surely it must be available in all READ operations. 
  * AFAIK OData v2 and AFAIK also the API gateway does not support the alternate key syntax yet. For these cases we probably need a 'fake local ID' so that the OID can really be used as key. But this is not my expertise. 
  * S/4 does not allow computed OIDs except for codes. The reason is that on a SELECT / READ, the whole OID table would be materialized before the comparison can be made. At leas that is the current restriction AFAIK in SADL.
* The OID annotation must have as parameter the name of the field that holds the OID value. This means that the field must exist in the schema so that it can exist in the payload right? Whether a value is computed or not by the application I would consider as an implementation detail that is internal to the application.

Bottom line: OID annotation must be given on the entity OID (head level) and the payload field must be there. 

I hope this helps.

Comment (ralfhandl) [reply tools/tests/annotations.openapi3.json:2902]: "Computed" here means server-generated and read-only. Whether that means "computed always on the fly" or "computed once and persisted" is an implementation detail.

If the OID is server-generated, it will not be part of the CREATE _request_ payload.

If the OID is immutable and can't change after entity creation, it will not be part of the UPDATE _request_ payload.

As a consequence the OID is only guaranteed to be part of _response_ payloads, aka the READ schema.

Comment (juergenheymann) [reply tools/tests/annotations.openapi3.json:2902]: Ah ... then I misunderstood. Then I agree with Ralf and Sebastian :-)

Comment (ralfhandl) [reply tools/tests/annotations.openapi3.json]: I agree but am not sure whether I can implement the property lookup (efficiently) with XSLT. Current shortcut is to only render `x-sap-odm-oid` for the READ structure and skip it for the other two.

Comment (swennemers) [reply tools/tests/annotations.openapi3.json]: okay, that means either not annotating create/update structures that might contain the OID field or annotating an OID field in the create/update structure that is not there?

Could you go 2 phases and remove the annotation in case the annotated property got removed from the create/update structure?

Comment (ralfhandl) [reply tools/tests/annotations.openapi3.json]: Phase 1 was to unconditionally remove `x-sap-odm-oid` from the CREATE and UPDATE structure because at that point during generation I don't have access to the properties.

That's already part of this PR.

Comment (juergenheymann) [reply tools/tests/annotations.openapi3.json]: Responding to Sebastians text:
1. YES  
2. YES, but: for update the OID can be used as identifier on the URL, right? 
3. YES. This will be the case for all kinds of OIDs used as references or for codes.
4. YES and I would add: "IF THERE ARE EVER cases, ..."

Comment (ralfhandl) [reply tools/tests/annotations.openapi3.json]: Note: 1. - 4. are Phase 2 and may be added in a future PR

**Review by juergenheymann — APPROVED** [review]
LGTM - as far as I understand the syntax

**Review by swennemers — APPROVED** [review]
Ok, then start with this version and have the create/update structures at a later point in time.

**Comment by pavelkornev** [comment]
LGTM.
